# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/ContentChild.h
# Commit: 22b6ef4463c6
# Full Hash: 22b6ef4463c68ed2a9a22a86cbb36956fc20627a
# Author: Jamie Nicol <jnicol@mozilla.com>
# Date: 2023-09-27 03:49:15
# Regressor Bug: 1850573
# File Overlap Count: 1
# Description:
#   Bug 1850573 - Create/destroy PerformanceHintSession in ContentChild based on process priority. r=smaug
#   
#   We use a PerformanceHintSession where supported to ensure the content
#   process main thread and stylo threads are scheduled at a high
#   priority. Previously we created the session when the
# ==============================================================================

diff -r 653daffe1a6b -r 22b6ef4463c6 dom/ipc/ContentChild.h
--- a/dom/ipc/ContentChild.h	Tue Sep 26 16:53:51 2023 +0300
+++ b/dom/ipc/ContentChild.h	Tue Sep 26 12:55:37 2023 +0000
@@ -13,6 +13,7 @@
 #include "mozilla/dom/PContentChild.h"
 #include "mozilla/dom/ProcessActor.h"
 #include "mozilla/dom/RemoteType.h"
+#include "mozilla/Hal.h"
 #include "mozilla/ipc/InputStreamUtils.h"
 #include "mozilla/ipc/ProtocolUtils.h"
 #include "mozilla/StaticPtr.h"
@@ -797,6 +798,18 @@
 
   hal::ProcessPriority GetProcessPriority() const { return mProcessPriority; }
 
+  hal::PerformanceHintSession* PerformanceHintSession() const {
+    return mPerformanceHintSession.get();
+  }
+
+  // Returns the target work duration for the PerformanceHintSession, based on
+  // the refresh interval. Estimate that we want the tick to complete in at most
+  // half of the refresh period. This is fairly arbitrary and can be tweaked
+  // later.
+  static TimeDuration GetPerformanceHintTarget(TimeDuration aRefreshInterval) {
+    return aRefreshInterval / int64_t(2);
+  }
+
  private:
   void AddProfileToProcessName(const nsACString& aProfile);
   mozilla::ipc::IPCResult RecvFlushFOGData(FlushFOGDataResolver&& aResolver);
@@ -813,6 +826,8 @@
       const Message& aMsg, UniquePtr<Message>& aReply) override;
 #endif
 
+  void ConfigureThreadPerformanceHints(const hal::ProcessPriority& aPriority);
+
   nsTArray<mozilla::UniquePtr<AlertObserver>> mAlertObservers;
   RefPtr<ConsoleListener> mConsoleListener;
 
@@ -883,6 +898,11 @@
   uint64_t mBrowsingContextFieldEpoch = 0;
 
   hal::ProcessPriority mProcessPriority = hal::PROCESS_PRIORITY_UNKNOWN;
+
+  // Session created when the process priority is FOREGROUND to ensure high
+  // priority scheduling of important threads. (Currently main thread and style
+  // threads.) The work duration is reported by the RefreshDriverTimer.
+  UniquePtr<hal::PerformanceHintSession> mPerformanceHintSession;
 };
 
 inline nsISupports* ToSupports(mozilla::dom::ContentChild* aContentChild) {